### ISSUE TITLE
Implicit error redirect

### ISSUE THREAD

**Issue by mortzdk** [bshaffer/oauth2-server-php#858]
In the [documentation](http://openid.net/specs/openid-connect-core-1_0.html#ImplicitAuthError) of OpenID Connect it states:

> If the End-User denies the request or the End-User authentication fails, the Authorization Server MUST return the error Authorization Response in the fragment component of the Redirection URI, as defined in 4.2.2.1 of OAuth 2.0 [RFC6749] and OAuth 2.0 Multiple Response Type Encoding Practices [OAuth.Responses], unless a different Response Mode was specified.

Currently the error response is creating a redirect using GET parameters instead of as fragment components when `allow_implicit` is set to `true`.

**Comment by mortzdk** [comment]
[Duplicate](https://github.com/bshaffer/oauth2-server-php/issues/837)

**Comment by markkimsal** [comment]
FYI, the OpenID spec has a "response_mode" parameter where the client can specify fragment or query parameters.  There's even a "form_post" response_mode where the AS generates an auto-submitting form (instead of Location based redirect) for the UA that POSTs to the client's redirect_uri.

To test out a $responseMode parameter that just handles queries and fragments I added
`protected $responseMode = 'fragment';`
to OAuth2/Controller/AuthorizeController.php  and then hacked up buildUri like so:

```
    private function buildUri($uri, $params)
    {
        $parse_url = parse_url($uri);

        // Add our params to the parsed uri
        // ...

        if ($this->responseMode == 'fragment') {
            if (!isset($parse_url['fragment'])) {
                $parse_url['fragment'] = '';
            }
            $parse_url['fragment'] .= $parse_url['query'];
            unset($parse_url['query']);
        }
        // Put humpty dumpty back together
        return
           //...
        ;
    }

```
... missing all the getters and setters of course.

The other solution would be tracking $responseMode all the way down to each responseType in the ResponseTypeInterface and then doing something like

```
    public function getAuthorizeResponse($params, $user_id = null)
    { 
        $result = $this->authCode->getAuthorizeResponse($params, $user_id);
        $id_token = $this->idToken->createIdToken($params['client_id'], $user_id, $params['nonce']);
        $result[1][  $this->responseMode  ]['id_token'] = $id_token;

        return $result;
    } 
```

But, response_mode is not really tied to a response type.  You can't mix and match query/fragment in one request.  It's part of the request and could be justified being handled by the controllers as much as justifying it being handled by each response type.

https://openid.net/specs/oauth-v2-multiple-response-types-1_0.html#ResponseModes

http://openid.net/specs/oauth-v2-form-post-response-mode-1_0-02.html

**Comment by bluebaroncanada** [comment]
Yeah.  I was reading that.  It says that it MUST return fragment unless the response mode is "specified".  Do you know if that means it should override the default?

**Comment by markkimsal** [comment]
I think OpenID uses fragment by default and OAuth uses query by default, but I have no docs to confirm that.

**Comment by bluebaroncanada** [comment]
It's the preference but the default can be set by the application.

**Comment by bluebaroncanada** [comment]
It does mention that, too, in the link in the first comment.
